[클린코드 리액트 2기 장태영] 페이먼츠 미션 Step 1

taeyoungs opened the pull request:
안녕하세요!
클린코드 리액트 2기 수강생 장태영입니다.
잘 부탁드립니다. 🙇‍♂️

더 늦기 전에 피드백을 한번 받고 싶어서 후다닥 달려왔습니다. 🫠

- [데모 페이지](https://react-payments-umber.vercel.app/)
- [Storybook](https://63f1c654e4f702113aa32a89-xhstpprgnz.chromatic.com/?path=/story/input--template)

## 필수 요구사항
- [x] Storybook 상호 작용 테스트
- [x] 재사용 가능한 Component 작성

## 카드 추가
- [x] <(뒤로가기) 버튼 클릭 시, 카드 목록 페이지로 이동한다.
- [x] 카드 번호를 입력 받을 수 있다.
   - [x] 카드 번호는 숫자만 입력가능하다.
   - [x] 카드 번호 4자리마다 -가 삽입된다.
   - [x] 카드 번호는 실시간으로 카드 UI에 반영된다.
   - [x] 카드 번호는 앞 8자리만 숫자로 보여지고, 나머지 숫자는 *로 보여진다.
- [x] 만료일을 입력 받을 수 있다.
   - [x] MM / YY 로 placeholder를 적용한다.
   - [x] 월, 년 사이에 자동으로 /가 삽입된다.
   - [x] 만료일은 실시간으로 카드 UI에 반영된다.
   - [x] 월은 1이상 12이하 숫자여야 한다.
- [x] 보안코드를 입력 받을 수 있다.
   - [x] 보안코드는 *으로 보여진다.
   - [x] 보안코드는 숫자만 입력가능하다.
- [x] 카드 비밀번호의 앞 2자리를 입력 받을 수 있다.
   - [x] 카드 비밀번호는 각 폼마다 한자리 숫자만 입력가능하다.
   - [x] 카드 번호 입력 시, *으로 보여진다.
- [x] 카드 소유자 이름을 입력 받을 수 있다.
   - [x] 이름은 30자리까지 입력할 수 있다.
   - [x] 이름 입력 폼 위에, 현재 입력 자릿수와 최대 입력 자릿수를 실시간으로 보여준다.
- [x] 카드 추가 완료시 카드 등록 완료 페이지로 이동한다.

## 애매했던 점

- 우선 CDD를 제대로 진행하진 못한 것 같습니다. 🥲 어떤 부분은 컴포넌트와 해당 컴포넌트를 위한 스토리북을 먼저 작성했고 어떤 부분은 도메인과 관련된 컴포넌트를 만들다가 필요하다 싶어서 다시 가서 만들기도 한 것 같습니다.
- 재사용 컴포넌트를 만든다고 만들어봤는데 어떤 컴포넌트는 이렇게 만들거면 만들지 않아도 됐었던 건가? 분리를 하긴 했지만 너무 이르게 분리를 해버린건지 .. 계산기 미션에서 아주 조금은 연습이 됐다고 생각했는데 .. 관심사 분리도 제대로 진행한건지 모르겠습니다. 🥲
- 'Storybook 상호 작용 테스트'은 성공적으로 수행한 것을 떠나서 나름대로 진행은 했기에 우선 체크했습니다! 😭

taeyoungs committed:
init: Project initialize (CRA + storybook)

taeyoungs committed:
chore: styles 폴더 이동 및 index.css import

taeyoungs committed:
feat: router 객체 및 Layout 컴포넌트 추가

taeyoungs committed:
feat: Card, CardList 컴포넌트 템플릿 추가

taeyoungs committed:
feat: 카드 추가 페이지 템플릿 추가

taeyoungs committed:
mod: Card 컴포넌트 업데이트

taeyoungs committed:
chore: 불필요한 컴포넌트 삭제

taeyoungs committed:
chore: 네이밍 수정 (serial => card)

taeyoungs committed:
feat: Input 컴포넌트 구현 및 헬퍼 타입 추가

taeyoungs committed:
feat: 카드 번호 관리를 위한 커스텀 훅 추가

taeyoungs committed:
feat: 카드 번호 컨테이너 컴포넌트 구현

taeyoungs committed:
mod: Input 컴포넌트 forwardRef 추가

taeyoungs committed:
feat: Dash 컴포넌트 구현

components/Dash/index.tsx
- Card 컴포넌트와 CardNumberContainer 컴포넌트에서 사용할
  '-'를 표현할 Dash 컴포넌트 구현

taeyoungs committed:
feat: Input focus 컨트롤을 위한 커스텀 훅 추가

taeyoungs committed:
mod: ref와 useInputFocusing을 이용한 focus 컨트롤

taeyoungs committed:
feat: Card 뒷 8자리 마스킹 처리 추가

taeyoungs committed:
mod: isNumber 함수를 유틸 함수로 분리

taeyoungs committed:
feat: 만료일 input 관련 커스텀 훅 구현

taeyoungs committed:
feat: 만료일 컴포넌트 ExpiredDateContainer 구현

taeyoungs committed:
mod: input placeholder 수정

taeyoungs committed:
mod: Card 컴포넌트 expiredDate 매핑

taeyoungs committed:
feat: InputProgress 컴포넌트 구현 및 스토리북 추가

taeyoungs committed:
chore: 불필요한 스토리북 파일 삭제

taeyoungs committed:
feat: CompanyColor 컴포넌트 구현 및 스토리북 추가

taeyoungs committed:
feat: Masking 컴포넌트 구현 및 스토리북 추가

taeyoungs committed:
mod: Masking 컴포넌트 CSS 업데이트

taeyoungs committed:
mod: '*' 문자열을 Masking 컴포넌트로 대체

taeyoungs committed:
mod: Input 컴포넌트 align 속성 추가

taeyoungs committed:
chore: components 폴더의 하위 폴더 구조 변경

taeyoungs committed:
chore: 공용 컴포넌트 스토리북 파일 위치 변경

taeyoungs committed:
feat: Dash 컴포넌트 스토리북 추가

taeyoungs committed:
feat: CardOwnerContainer 및 useCardOwner 구현

taeyoungs committed:
mod: Input, Card 컴포넌트 CSS 업데이트

taeyoungs committed:
mod: Input 컴포넌트 CSS 업데이트

taeyoungs committed:
mod: Input size 속성 추가 및 스토리북 수정

taeyoungs committed:
feat: SecretCodeContainer(보안코드) 컴포넌트 구현

taeyoungs committed:
mod: Masking 컴포넌트 CSS 업데이트

taeyoungs committed:
feat: CardPasswordContainer(비밀번호) 컴포넌트 구현

taeyoungs committed:
feat: CardCompanyPicker(카드사 선택) 컴포넌트 구현

taeyoungs committed:
chore: CompanyColor, Card 컴포넌트 CSS 업데이트

taeyoungs committed:
mod: 카드 추가 페이지 CardCompanyPicker 추가

taeyoungs committed:
feat: 카드 등록 확인 페이지 템플릿 추가

taeyoungs committed:
chore: 불필요한 스토리북 파일 삭제

taeyoungs committed:
feat: Button 컴포넌트 구현 및 스토리북 추가

taeyoungs committed:
feat: Header 컴포넌트 구현 및 스토리북 추가

taeyoungs committed:
feat: 페이지 이동 버튼 추가

taeyoungs committed:
chore: 사용되지 않는 masking 함수 삭제

taeyoungs committed:
chore: chromatic deploy

taeyoungs committed:
chore: chromatic deploy

InSeong-So commented on package.json:
```diff
@@ -0,0 +1,76 @@
+    "react-dom": "^18.2.0",
+    "react-router-dom": "^6.8.1",
+    "react-scripts": "5.0.1",
+    "typescript": "^4.4.2",
```
typescript도 devDependencies에 위치하는 것이 맞겠네요😊

taeyoungs replied:
넵 ! devDependencies로 옮겨놓도록 하겠습니다. :)

taeyoungs commented:
안녕하세요 인성님!

계속 고민했던 것도 여쭤보려고 코멘트 하나만 달아놓겠습니다!

스토리북을 추가해주는 컴포넌트들은 보통 도메인과 무관하고 재사용이 가능한 컴포넌트들일까요??
도메인과 관련된 컴포넌트들도 스토리북을 통한 상호 작용 테스트를 위해 스토리북을 추가해주는지 궁금합니다.
스토리북 쇼케이스를 보면 도메인과 무관하고 재사용이 가능한 컴포넌트들인데 이게 공개되어 있는 거라 그런건지 아니면
원래 도메인과 무관하고 재사용이 가능한 컴포넌트들 위주로 작성되고 그 외에 도메인과 관련된 컴포넌트들은 다른 방식으로 진행하는건지 .. 잘 모르겠어서 댓글을 남깁니다! (당연한 이야기일까요.. 🥲 당연히 케이스마다 다를 것 같기도 하고 .. 😇)

InSeong-So commented on src/App.tsx:
```diff
@@ -0,0 +1,25 @@
+import { AddingCard, ConfirmCard, HoldingCards } from 'pages';
+import { Layout } from 'components/common';
+
+const router = createBrowserRouter(
+  createRoutesFromElements(
+    <Route path="/" element={<Layout />}>
+      <Route index element={<HoldingCards />} />
+      <Route path="new" element={<AddingCard />} />
+      <Route path="confirm" element={<ConfirmCard />} />
+    </Route>
+  )
+);
+
+function App() {
+  return <RouterProvider router={router} />;
+}
```
저는 App 파일을 Context나 Wrapper 컴포넌트를 둘 때 활용하는 편입니다. Provider, Suspense, ErrorBoundary... 그럴 때 App에는 필요 이상의 초기화/할당 로직이 많아지게 되어서 필연적으로 잘게 분리하는 편이에요.

router 객체는 다른 파일로 빼는 것이 좀 더 좋지 않을까? 하는 생각인데 어떻게 생각하시나요?
또한 path의 값을 타입이나 상수로 선언해두면 다른 컴포넌트나 페이지, 로직에서도 접근하기 좋을 것 같네요!

taeyoungs replied:
좋은 방법인 것 같습니다!
router 객체를 분리하고 path 값도 타입이나 상수로 선언해두면 router에 관련된 코드만 모아둘 수 있는 장점도 있겠네요! 👍

InSeong-So commented on src/components/domain/Card/index.tsx:
```diff
@@ -0,0 +1,70 @@
+            >
+              {name}
+            </span>
+            <span className="card-text">
+              {month} / {year}
+            </span>
```
`월, 년 사이에 자동으로 /가 삽입된다.` 이 요구사항에 위배되는 코드 같네요!

taeyoungs replied:
앗 수정하도록 하겠습니다.

InSeong-So commented on src/components/common/CompanyColor/index.tsx:
```diff
@@ -0,0 +1,49 @@
+  onClick: React.MouseEventHandler<HTMLDivElement>;
+}
+
+function CompanyColor({ company, onClick }: CompanyColorProps) {
+  return (
+    <div
```
컴포넌트-도메인에 있어야 하지 않을까요?🤔

taeyoungs replied:
쓰읍 .. 애매하다고 생각했는데 역시 지금 제가 domain과 common으로 나눠놓은 관점에서는 components > domain 하위에 있는게 맞는 것 같습니다. 옮겨놓도록 하겠습니다. :)

InSeong-So commented on src/components/common/Button/index.tsx:
```diff
@@ -0,0 +1,41 @@
+    <button
+      {...props}
+      className={cx(
+        css`
+          font-weight: ${weight};
+          font-size: ${typeof fontSize === 'number' ? fontSize + 'px' : fontSize};
+          color: ${color};
+          background-color: transparent;
+          border: none;
+          box-sizing: border-box;
+        `,
```
전체적으로 컴포넌트에 이런 패턴이 있는데, 이렇게 스타일 또는 클래스명을 정의하는 경우 변수/함수로 외부에 분리하여 가져가주는 것이 가독성에 도움을 줍니다! 본 컴포넌트의 라인 수가 길어져도 태그를 읽기 난해해지거든요😅

taeyoungs replied:
스타일이 아닌 다른 곳에선 이런 부분들을 고려했었던 것 같은데 여기선 미처 고려하지 못했던 것 같습니다.
분리해서 가독성에 조금 더 도움이 되도록 만들어보겠습니다.

InSeong-So commented on src/components/common/Dash/index.tsx:
```diff
@@ -0,0 +1,22 @@
+function Dash({ visible = false, width = 6, height = 3 }: DashProps) {
+  return (
+    <span
+      className={css`
+        display: inline-block;
+        padding: 0 ${width / 2}px;
+        height: ${height}px;
+        background-color: ${visible ? '#000' : 'transparent'};
+      `}
```
공통(common)으로 묶여있는 컴포넌트들은 props를 다룰 때 일관되게 적용하도록 작성해주시면 좋습니다. 현재는 number 타입만 받네요! 다른 컴포넌트에서는 CSSProperties의 특정 키-값을 받아서 삼항식으로 처리해주셨는데, 저는 후자를 일관되게 가져가면 너무 좋을 것 같습니다🥳

taeyoungs replied:
'CSSProperties의 특정 키-값을 받는' 형태로 일관성 있게 가져가보도록 하겠습니다!

InSeong-So commented on src/components/common/Header/index.tsx:
```diff
@@ -0,0 +1,36 @@
+        align-items: center;
+        gap: 18px;
+      `}
+    >
+      {prefix}
+      <span
```
ReactNode가 들어오는 자리가 prefix... 흠, 좀 더 직관적인 네이밍이 있으면 좋겠네요! 접미어의 용도로 많이 받아들여지기 때문에, 단순 텍스트가 아니라 컴포넌트가 위치하게 되면 더 직관적으로 가져갈 필요가 있지 않을까 싶습니다🥹

taeyoungs replied:
으음 .. 직관적이라는 키워드에 맞춰서 수정을 해보겠습니다. 👍

InSeong-So commented on src/components/common/Input/index.tsx:
```diff
@@ -0,0 +1,53 @@
+
+export type ButtonSize = 'small' | 'medium' | 'large' | 'full';
+
+const inputSize = {
+  small: '15%',
+  medium: '25%',
+  large: '50%',
+  full: '100%',
+};
```
다른 컴포넌트에서는 외부에 위치한 상수를 UPPER_CASE로 정의하셨던데, 이것도 그렇게 작성해주시면 더 좋겠네요!

taeyoungs replied:
넵넵 ! 일관성있게 가져가겠습니다.

InSeong-So commented on src/components/common/Input/index.tsx:
```diff
@@ -0,0 +1,53 @@
+import { CSSProperties, forwardRef } from 'react';
+
+import type { CombineElementProps } from 'types/utils';
+
+export type ButtonSize = 'small' | 'medium' | 'large' | 'full';
+
```
그리고 내부에서 Button으로 쓰이는 곳이 없는데 Button이라는 이름은 난해하지 않을까요?!

taeyoungs replied:
엇 ! 이건 제 실수 같습니다. 왜 이렇게 작성했는지 모르겠습니다. 😇 수정하도록 하겠습니다!

InSeong-So commented on src/components/common/InputProgress/index.tsx:
```diff
@@ -0,0 +1,30 @@
+  current: number;
+  max: number;
+}
+
+function InputProgress({ current, max }: InputProgressProps) {
+  return (
```
이 컴포넌트는 Input과 밀접한 연관이 있기 때문에, Input 컴포넌트 하위에 정의하거나 조건문을 통해 핸들링 하도록 하면 어떨까요?

taeyoungs replied:
Input 컴포넌트에서 이미 InputProgress에서 필요한 값들(maxLength, value의 길이)을 이미 알고 있기도 하네요.
이를 이용해서 하위에서 정의하고 핸들링 해보도록 하겠습니다. 🚀

InSeong-So commented on src/components/common/Layout/index.tsx:
```diff
@@ -0,0 +1,13 @@
+import { Outlet } from 'react-router-dom';
+
+function Layout() {
+  return (
```
이런 코드는 재사용이 가능한 컴포넌트 폴더-"공통"-보다는 최상단에서 "한 번" 호출되는 형태이기 때문에 컴포넌트 위치보다 Base 경로 혹은 Router를 정의한 파일에 같이 선언해주면 좋을 것 같은데, 어떻게 생각하시나요?!

taeyoungs replied:
저도 components - common 하위에 있는게 과연 맞는걸까에 대한 고민을 했었는데 어딘가에 Layout이라는 이름으로 분리해놓고 싶어서 이렇게 진행했던 것 같습니다!
위에서 나온 피드백(router 분리)을 반영하면서 이 부분도 같이 적용하면 좋을 것 같습니다. 👍

InSeong-So commented on src/components/common/Masking/index.tsx:
```diff
@@ -0,0 +1,37 @@
+  gap?: number;
+  color?: CSSProperties['color'];
+}
+
+function Masking({ count, width = 4, height = 4, gap = 2, color = '#000' }: MaskingProps) {
+  return (
```
요것도 도메인이 옳지 않을까... 싶네요. 공통 컴포넌트로써 사용될 니즈가 잘 안보여서요😅

taeyoungs replied:
역시 그런가요 .. 이런 부분이 아직 감이 오지 않는 것 같습니다. UI 컴포넌트 라이브러리들에서 자주 나오는 키워드들(Input, Button, Select 등) 정도는 되어야 공통 컴포넌트로써의 니즈가 발생한다고 볼 수 있는걸까요 ?? 🤔

InSeong-So commented on src/components/common/Input/index.tsx:
```diff
@@ -0,0 +1,53 @@
+interface InputBaseProps {
+  textAlign?: CSSProperties['textAlign'];
+  size?: ButtonSize;
+}
+export type InputProps = CombineElementProps<'input', InputBaseProps>;
+
```
태영님이 작성한 컨벤션이라면 이 코드도 interface가 더 옳을 것 같은데... type으로 선언해주셨군요? 제가 잘못 일반화한 것일까요🥲

taeyoungs replied:
아뇨 ! 맞습니다. 작성하면서 왔다리 갔다리 했던 것 같네요. 일관성있게 interface로 가져가보겠습니다. 🥹

InSeong-So commented on src/components/domain/Card/index.tsx:
```diff
@@ -0,0 +1,70 @@
+import { theme } from 'constants/colors';
+import type { CardCompany, ExpiredDate, CardNumber } from 'types/card';
+
+interface CardProps {
+  company: CardCompany | null;
+  name: string;
+  cardNumber: CardNumber;
+  expiredDate: ExpiredDate;
+}
```
company를 null이 아닌 default value로 선언해두면 아래처럼 3항식을 쓰지 않아도 되니, 관리 포인트가 줄겠죠! null을 빼보면 어떨까요?

taeyoungs replied:
음 ~ 넵넵 ! 🚀 좋습니다. :)

InSeong-So commented on src/components/domain/Card/index.tsx:
```diff
@@ -0,0 +1,70 @@
+          <div className="small-card__chip"></div>
+        </div>
+        <div className="card-bottom">
+          <div className="card-bottom__number">
+            <span className="card-number">{num1}</span>
+            <Dash visible={num1.length === MAX_LENGTH.CARD_NUMBER} width={4} height={1} />
+            <span className="card-number">{num2}</span>
+            <Dash visible={num2.length === MAX_LENGTH.CARD_NUMBER} width={4} height={1} />
+            <Masking count={num3.length} />
+            <Dash visible={num3.length === MAX_LENGTH.CARD_NUMBER} width={4} height={1} />
+            <Masking count={num4.length} />
```
뭔가 아쉽지... 했는데...! Dash 컴포넌트 내부에서 visible 여부에 따라 그려줄고 말고를 결정할 게 아니라, 현재 주입하는 조건문을 바깥에서 선언해주면 더 좋지 않을까요?
Dash 컴포넌트가 외부 로직을 "주입 받아서" 자체적으로 보여줄 지 여부를 결정하는 것이 조금 불필요하게 느껴지네요!

```jsx
{num1.length === MAX_LENGTH.CARD_NUMBER && <Dash width={4} height={1} />}
```

taeyoungs replied:
음 .. 지금 생각해보니 그렇네요. 굳이 외부에서 주입받을 필요가 없는 것 같습니다. 반영해보도록 하겠습니다! 👍

InSeong-So commented on src/components/domain/CardCompanyPicker/index.tsx:
```diff
@@ -0,0 +1,31 @@
+            `
+          )}
+        >
+          {Object.values(CardCompany).map((company) => (
+            <CompanyColor key={company} company={company} onClick={() => updateCompany(company)} />
+          ))}
```
`Object.values(CardCompany)`도 상단에 상수로 선언해두는 것이 좋겠어요!

InSeong-So commented on src/components/domain/CardNumberContainer/index.tsx:
```diff
@@ -0,0 +1,77 @@
+function CardNumberContainer({ cardNumber, handleChangeCardNumber }: CardNumberContainerProps) {
+  const { num1, num2, num3, num4 } = cardNumber;
+
+  const num1Ref = useRef<HTMLInputElement>(null);
+  const num2Ref = useRef<HTMLInputElement>(null);
+  const num3Ref = useRef<HTMLInputElement>(null);
+  const num4Ref = useRef<HTMLInputElement>(null);
```
이렇게도 될 수 있을 것 같아요! ref 객체가 많아질 때 자주 사용한답니다😊

```suggestion
const numRefs = Array.from(4).map(createRef<HTMLInputElement>);
```

taeyoungs replied:
오오 좋은 방법 감사합니다!

InSeong-So commented on src/hooks/useInputFocusing.ts:
```diff
@@ -0,0 +1,36 @@
+
+const useInputFocusing = ({ refs, deps, maxLength }: UseInputFocusingProps) => {
+  useEffect(() => {
+    const focusedInput = refs.find((ref) => document.activeElement === ref.current);
+
```
`(ref) => document.activeElement === ref.current` 콜백이 중복되어 발생하는게 조금 걸리네요. 유틸 함수로 분리해볼까요? findActiveElement 등으로 말이죠!

```suggestion
const findActiveElement = (ref) => document.activeElement === ref.current;

const focusedInput = refs.find(findActiveElement);
const focusedInputIndex = refs.findIndex(findActiveElement);
```

아니라면... findIndex 함수만 사용하여 `refs[index]` 접근하는 것도 방법입니다.

```suggestion
const focusedInputIndex = refs.findIndex((ref) => document.activeElement === ref.current);

if (focusedInputIndex === -1) return;

const focusedInput = refs[focusedInputIndex];
```

taeyoungs replied:
둘 중 하나의 방법으로 선택해서 수정해보겠습니다!

InSeong-So commented on src/hooks/useInputFocusing.ts:
```diff
@@ -0,0 +1,36 @@
+    const value = focusedInput.current?.value;
+    const focusedInputIndex = refs.findIndex((ref) => document.activeElement === ref.current);
+
+    if (value?.length === maxLength) {
+      const nextInput = refs[focusedInputIndex + 1];
+
+      nextInput && nextInput.current?.focus();
+    }
+
+    if (value?.length === 0) {
+      const previousInput = refs[focusedInputIndex - 1];
+
+      previousInput && previousInput.current?.focus();
+    }
```
이렇게 되면, 예외 상황이 하나 존재하는데요!

maxLength가 0이 들어오면 어떻게 될까요?!

InSeong-So commented on src/components/domain/CardPasswordContainer/index.tsx:
```diff
@@ -0,0 +1,73 @@
+    maxLength: MAX_LENGTH.PASSWORD,
+  });
+
+  const handleChangePassword: React.ChangeEventHandler<HTMLInputElement> = (e) => {
+    const { name, value } = e.currentTarget;
+
+    if (value !== '' && !isNumber(value)) {
+      return;
+    }
+
+    setPassword((prevState) => ({ ...prevState, [name]: value }));
+  };
```
여기는 아직 훅으로 분리하지 않은 걸까요? AddingCard 하위의 훅들과 똑같아 보이는 로직인데, 이것만 여기에 존재하고 있군요!

InSeong-So commented on src/pages/AddingCard/hooks/useCardOwner.ts:
```diff
@@ -0,0 +1,18 @@
+import { useState } from 'react';
+
+const useCardOwner = () => {
+  const [cardOwner, setCardOwner] = useState('');
+
+  const handleChangeCardOwner: React.ChangeEventHandler<HTMLInputElement> = (e) => {
+    const { value } = e.target;
+
+    setCardOwner(value);
+  };
+
+  return {
+    cardOwner,
+    handleChangeCardOwner,
+  };
+};
```
그냥 컴포넌트에서 사용하는 useState - handler 래핑 객체라는 느낌이🥲 setter만 은닉화된 상황인데, 이렇게 되면 ref 객체는 많이 사용하게 되면서 제어 컴포넌트를 남용하게 되지 않을까 하는 우려가 생기네요!

InSeong-So commented on src/pages/AddingCard/hooks/useExpiredDate.ts:
```diff
@@ -0,0 +1,41 @@
+    if (value !== '' && !isNumber(value)) {
+      return;
+    }
+
+    if (name === 'month' && isOutOfRangeMonth(value)) {
+      return;
```
'month'는 분명 input 태그에 name에 매직 스트링으로 주입하는 값일텐데... 상수로 관리하면 좋지 않을까요?!

InSeong-So commented on src/pages/AddingCard/hooks/useExpiredDate.ts:
```diff
@@ -0,0 +1,41 @@
+  return { expiredDate, handleChangeExpiredDate };
+};
+
+const isOutOfRangeMonth = (month: string) => {
+  if (month.length === 1) {
+    return month !== '0' && month !== '1';
+  }
+
+  if (month.length === 2) {
+    return +month <= 0 || +month > 12;
+  }
+};
```
이것도 유틸리티 함수로 정의해두면 좋지 않을까 싶습니다😊 Month가 유효한 범위인지 체크하는 함수는 어느 도메인에서도 잘 사용할 수 있을 것 같으니까요!

추가적으로 0, 1, 2, 12 등도 의미하는 바를 직관적으로 보여주기 위해 상수화 해두면 좋겠습니다!

taeyoungs replied:
'Month가 유효한 범위인지 체크하는 함수' .. 그렇네요!
유틸 함수로 분리하고 상수 처리도 해놓겠습니다. 👍

InSeong-So commented on src/pages/AddingCard/index.tsx:
```diff
@@ -0,0 +1,67 @@
+  const { cardOwner, handleChangeCardOwner } = useCardOwner();
+  const { open, company, updateCompany, openPicker } = useCompanyPicker();
+
+  const handleSubmit: React.FormEventHandler<HTMLFormElement> = (e) => {
+    e.preventDefault();
+
+    navigate('/confirm');
+  };
```
이런 것도 useRouter 같은 이름으로 훅을 만들어서 사용하면 위와 어느정도 통일성을 갖추면서, 좀 더 직관적이겠네요!

InSeong-So commented on src/types/card.ts:
```diff
@@ -0,0 +1,27 @@
+export enum CardCompany {
+  Shinhan = '신한',
+  Hana = '하나',
+  Woori = '우리',
+  Kb = '국민',
+  Lotte = '롯데',
+  Nh = '농협',
+  Samsung = '삼성',
+  Hyundai = '현대',
+}
```
enum을 활용해주셨네요! 평소에도 자주 사용하는 타입스크립트 문법이신가요?👀

taeyoungs replied:
아뇨! 계산기 미션 때 준일님의 피드백이 있었는데 이를 떠올리며 사용해봤습니다.
사실 아직 어떤 때에 enum을 사용하면 좋을지에 대해 조사가 제대로 이루어지지 않았는데 이르게 사용한 감도 있네요. 🥲
피드백을 반영할 때 이에 대한 조사도 한번 해서 정리해보겠습니다. :)

InSeong-So commented on src/types/utils.ts:
```diff
@@ -0,0 +1,5 @@
+import { ComponentProps, ElementType } from 'react';
+
+type Combine<T, K> = T & Omit<K, keyof T>;
+
+export type CombineElementProps<E extends ElementType, P = unknown> = Combine<P, ComponentProps<E>>;
```
👍

InSeong-So reviewed (changes requested):
태영님, 안녕하세요! 리뷰어 소인성입니다🙇‍♂️

어려운 요구사항임에도 불구하고 잘 개발해주셨어요! 컴포넌트가 상당히 잘게 쪼개져 있어서 어떤 맥락에서 폴더를 분리하고 domain과 common을 나누셨는지 궁금도 했습니다.

> 우선 CDD를 제대로 진행하진 못한 것 같습니다. 🥲 어떤 부분은 컴포넌트와 해당 컴포넌트를 위한 스토리북을 먼저 작성했고 어떤 부분은 도메인과 관련된 컴포넌트를 만들다가 필요하다 싶어서 다시 가서 만들기도 한 것 같습니다.
- 항상 고민이죠😅 처음 개발에 착수할 때 설계를 정확하게 거치지 않으면 중구난방으로 컴포넌트가 생기는데 시간은 없고... 하지만 잘해주셨습니다! 스토리북은 많이 사용해봐야 자신만의 보일러 플레이트가 생기기도 하니까요.

> 재사용 컴포넌트를 만든다고 만들어봤는데 어떤 컴포넌트는 이렇게 만들거면 만들지 않아도 됐었던 건가? 분리를 하긴 했지만 너무 이르게 분리를 해버린건지 .. 계산기 미션에서 아주 조금은 연습이 됐다고 생각했는데 .. 관심사 분리도 제대로 진행한건지 모르겠습니다. 🥲
- 코멘트를 남긴 것도, 아닌 것도 있습니다! 요청하신대로 빠른 피드백을 한 번 적용해보시는 게 도움이 될 것 같아요.

고생 많으셨어요!
남긴 코멘트 확인해주시고, 수정 뒤 다시 재요청 주시면 감사하겠습니다😊
+ 추가적으로 파일 전체적으로 EOL이 존재하는데, 일관성 있게 제거해주면 어떨까요?!

taeyoungs committed:
mod: router 객체 분리 및 라우팅 경로 상수화

taeyoungs committed:
chore: typescript를 devDependencies로 이동

taeyoungs committed:
chore: EOL 제거

taeyoungs committed:
chore: Masking, CompanyColor를 도메인 폴더 하위로

taeyoungs committed:
chore: import 경로 오류 수정

taeyoungs committed:
mod: Dash 컴포넌트 업데이트

components/common/Dash
- visible 속성 삭제 후 외부에서 Dash 컴포넌트를 조건부 렌더링하는
  방식으로 변경
- width 속성을 받아 padding에 바인딩해주는 상태가 혼란을 줄 수도
  있다고 생각하여 after 선택자를 이용한 방식으로 변경

taeyoungs committed:
mod: Card 컴포넌트 업데이트

componsnts/domain/Card
AddingCard/hooks/useCompanyPicker.ts
- company 상태값에 대한 초기값을 부여하고 null을 제거하여
  Card 컴포넌트 내에서 null 값으로 인한 삼항 연산자를 제거

taeyoungs committed:
chore: pixelOf 유틸 함수 분리

taeyoungs committed:
mod: 삼항 연산자를 pixelOf 함수로 대체

taeyoungs committed:
mod: 카드 추가 페이지 내 input name 문자열 상수화

taeyoungs committed:
mod: Input 컴포넌트 업데이트

taeyoungs committed:
chore: .root에 대한 CSS 업데이트

taeyoungs committed:
mod: Label 컴포넌트 활용

taeyoungs committed:
feat: 여러 개의 Ref를 생성해서 반환하는 훅 추가

taeyoungs committed:
mod: 보안코드 상태값도 커스텀 훅으로 분리

taeyoungs committed:
mod: Object.values(CardCompany) 상수 처리

taeyoungs committed:
mod: isOutOfRangeMonth 함수를 유틸 함수로 분리

taeyoungs committed:
chore: 오타 수정

taeyoungs committed:
mod: useInputFocusing 내 중복 콜백 함수 분리

taeyoungs committed:
chore: chromatic deploy

taeyoungs committed:
mod: Header 컴포넌트 prefix 속성 네이밍 변경

taeyoungs committed:
chore: 오타 수정

taeyoungs commented:
안녕하세요. 인성님! 조금 늦었습니다. 🥲

> 여기는 아직 훅으로 분리하지 않은 걸까요? AddingCard 하위의 훅들과 똑같아 보이는 로직인데, 이것만 여기에 존재하고 있군요!

훅으로 분리를 진행했습니다! AddingCard 하위에 있는 이상 일관적으로 분리해주는 것이 좋을 것 같아서 이 부분도 동일하게 진행했습니다.

> 그냥 컴포넌트에서 사용하는 useState - handler 래핑 객체라는 느낌이🥲 setter만 은닉화된 상황인데, 이렇게 되면 ref 객체는 많이 사용하게 되면서 제어 컴포넌트를 남용하게 되지 않을까 하는 우려가 생기네요!

Card 컴포넌트에게도 cardOwner 값을 전달하려다 보니 지금과 같은 형태가 됐습니다. 저도 'useState - handler 래핑 객체'라는 느낌을 지울 수가 없는데 현재 구조에선 더 좋은 방법이 떠오르지 않습니다. 🫠
상위 컴포넌트에서 전체적으로 Form 관련 상태를 관리하고 이를 Context API로 전달하는 형태면 'useState - handler 래핑 객체'와 같은 크게 의미가 있지 않은 부분을 제거할 수 있을까요 ..

> 이렇게 되면, 예외 상황이 하나 존재하는데요! maxLength가 0이 들어오면 어떻게 될까요?!

우선 maxLength가 0일 때 따로 동작하지 않게 만들어 놨습니다. 이 부분도 더 좋은 방법이 떠오르지가 않네요. 🥲

> 이 컴포넌트는 Input과 밀접한 연관이 있기 때문에, Input 컴포넌트 하위에 정의하거나 조건문을 통해 핸들링 하도록 하면 어떨까요?

InputProgress를 Input 하위로 옮김과 동시에 이것저것 수정을 하다 보니 생각보니 피드백을 반영하기 전 Input과 달라진 것 같습니다 .. 😇 
스토리북으로 먼저 작업하고 반영을 하긴 했지만 .. 더 복잡해진 것 같네요.

## 그 외

### 컴포넌트 domain과 common으로 분리한 이유

단순하게 재사용 가능한 공용 컴포넌트와 도메인 관련된 컴포넌트를 분리하고 싶었습니다. 
한 공간 안에 이 둘이 섞여 있으면 스스로 헷갈릴 것 같아서요!
기존에 이런 식(common과 domain)으로 분류해본 적이 없긴 한데 domain이라는 키워드에 꽂혀서 분류하게 된 것 같습니다.

### 스토리북 관련

스토리북을 추가해주는 컴포넌트들은 보통 도메인과 무관하고 재사용이 가능한 컴포넌트들일까요??
도메인과 관련된 컴포넌트들도 스토리북을 통한 상호 작용 테스트를 위해 스토리북을 추가해주는지 궁금합니다.
스토리북 쇼케이스를 보면 도메인과 무관하고 재사용이 가능한 컴포넌트들인데 이게 공개되어 있는 거라 그런건지 아니면 원래 도메인과 무관하고 재사용이 가능한 컴포넌트들 위주로 작성되고 그 외에 도메인과 관련된 컴포넌트들은 다른 방식으로 진행하는건지 .. 잘 모르겠어서요!
(당연한 이야기일까요.. 🥲 당연히 케이스마다 다를 것 같기도 하고 .. 😇)

### Enum과 Uppercase 상수

인성님은 상수 처리를 하실 때 Enum과 Uppercase 객체 둘 중에 선호하시는 방법이 있나요? 물론 때에 따라 다르겠지만 궁금해서 여쭤봅니다!

장단점은 다음과 같이 정리해봤습니다.


**Enum**
장점:

- 가독성이 높아집니다. 이유는 enum으로 값을 작성하면 값의 의미를 좀 더 명확하게 전달할 수 있습니다.
- 중복 코드가 줄어듭니다. 여러 번 사용되는 상수 값을 enum으로 작성하면 중복 코드를 줄일 수 있습니다.

단점:

- enum 값을 사용할 때, 일부 상황에서 타입 검사가 제한될 수 있습니다.

**Uppercase**

장점:

- 코드가 더욱 명시적이고 직관적입니다.
- 다른 객체와 마찬가지로 타입 검사가 가능합니다.

단점:

- 상수 값이 중복될 가능성이 높기 때문에, 이를 방지하기 위해 중복된 상수가 있는지 검사할 필요가 있습니다.

## 피드백 체크리스트

- [x]  router 객체 분리
- [x]  path 상수 및 재사용
- [x]  EOL 제거
- [x]  'CSSProperties의 특정 키-값을 받는' 형태로 일관성 있게 가져가기
- [x]  스타일에 인라인으로 작성된 삼항 연산자를 인라인이 아닌 변수나 함수로 분리하여 가독성을 높이기
- [x]  ComanyColor 컴포넌트를 components > domain 하위로 이동
- [x]  `월, 년 사이에 자동으로 /가 삽입된다.` 요구 사항 반영
- [x]  Dash 컴포넌트의 visible 속성 제거 및 외부에서 렌더링 여부를 결정하게끔 변경
- [x]  Card 컴포넌트에서 company 속성에 null 타입 대신 default value를 설정
- [x]  interface로 컨벤션을 일관적으로 가져가기
- [x]  Masking 컴포넌트를 components > common 하위로 이동
- [x]  재사용될 가능성이 없는 Layout 컴포넌트 components > common이 아니라 라우터랑 묶어주기
- [x]  InputProgress를 Input 컴포넌트 하위로 옮겨서 핸들링 해보기
- [x]  Input 컴포넌트 내에 존재하는 ButtonSize 타입 네이밍 수정
- [x]  'month' 문자열 상수 처리
- [x]  useInputFocusing 중복 로직 최소화하기
- [x]  useInputFocusing에서 maxLength가 0인 케이스 대응
- [x]  `Array.from(4).map(createRef<HTMLInputElement>);`
- [x]  `Object.values(CardCompany)` 상수화
- [x]  `isOutOfRangeMonth` 유틸 함수 분리 및 관련 숫자, 문자 상수 처리

InSeong-So commented on src/utils.ts:
```diff
@@ -0,0 +1,17 @@
+export const isNumber = (value: string) => /^\d+$/.test(value);
+
+export const pixelOf = (value: string | number) =>
+  typeof value === 'number' ? `${value}px` : value;
```
이 함수의 위험한 점은... `%, em, rem` 단위가 들어올 때 무슨 일이 일어날까입니다 ㄷㄷ

taeyoungs replied:
> 타입스크립트에는 타입 공간과 값 공간이라는 게 존재합니다. 값 공간은 자바스크립트 코드가 존재하며, 타입 공간은 타입만 위치하게 돼요. 그런데 두 곳 모두 존재하는 코드가 있습니다. typeof, class, enum 등이 대표적이에요. 더 이질적인 것은 enum 입니다. 타입스크립트에서 파생된 문법임에도 불구하고 자바스크립트 코드로 변환되는 것은... "컴파일 레벨에서 상수로 변경되는 것인데 그럴거면 enum이 아니라 상수를 쓰겠다" 입니다.
> 또한 Enum은 근본적인 문제가 있습니다. 인덱스와 키로 값에 접근할 수 있는 장점이 있지만, 강타입 언어에서 강제하는 indeOfRange 등의 에러는 타입스크립트의 Enum에 존재하지 않습니다. 이런 이유로, 저는 일절 사용하지 않아요.

자세한 답변 정말 감사합니다! 🙇‍♂️

InSeong-So reviewed (approved):
안녕하세요, 태영님!
피드백을 적용하려고 많은 노력을 하셨네요. 정말 고생 많으셨습니다!
확실히 코드 가독성이 많이 올라갔고, 리팩토링도 잘 이루어진 것 같아요.

질문에 답변드리자면☺

> 스토리북 관련
- 컴포넌트 주도 개발은 재사용 가능한 여러 컴포넌트가 잘 동작하는지 테스트와 함께 무결성을 검사한 뒤, 검증이 완료되면 컴포넌트와 컴포넌트를 합쳐 새로운 컴포넌트를 만들어 다시 테스트하며 앱을 구성하는 기법입니다. 스토리북은 정적인 UI 뿐만 아니라 props 및 사용자 인터렉션을 확인할 수 있어 매우 좋은 프레임워크죠..
- 즉, 유틸리티적인 원자 단위의 재사용 가능한 컴포넌트 뿐만 아니라 도메인 컴포넌트 또한 같이 작업할 수 있습니다🤗 보통 아토믹 디자인 패턴을 이 시점에 많이 언급하는데요, Organisms, Templates, Pages 를 작성하는 시점에는 이미 도메인 로직이 많이 들어가 있는 컴포넌트를 체크하게 될 겁니다.

> Enum과 Uppercase 상수
- 저는 Enum의 장점과 상관 없이 사용하지 않습니다🤗 본질적인 문제로요.
- 타입스크립트에는 타입 공간과 값 공간이라는 게 존재합니다. 값 공간은 자바스크립트 코드가 존재하며, 타입 공간은 타입만 위치하게 돼요. 그런데 두 곳 모두 존재하는 코드가 있습니다. typeof, class, enum 등이 대표적이에요. 더 이질적인 것은 enum 입니다. 타입스크립트에서 파생된 문법임에도 불구하고 자바스크립트 코드로 변환되는 것은... "컴파일 레벨에서 상수로 변경되는 것인데 그럴거면 enum이 아니라 상수를 쓰겠다" 입니다.
- 또한 Enum은 근본적인 문제가 있습니다. 인덱스와 키로 값에 접근할 수 있는 장점이 있지만, 강타입 언어에서 강제하는 indeOfRange 등의 에러는 타입스크립트의 Enum에 존재하지 않습니다. 이런 이유로, 저는 일절 사용하지 않아요.
- Uppercase로 선언한 상수의 장점은... `다른 객체와 마찬가지로 타입 검사가 가능합니다.` 이건 아닌 것 같네요😅 그냥 타입스크립트의 추론 기능이 잘 동작했기 때문이고, 실제 장점은 가독성과 다른 변수들과의 구분 정도입니다.
- 단점이라고 언급한 내용의 경우, 파일을 구조화하여 가져간다면 발생할 여지가 없기 때문에 `중복`이 문제가 되진 않을 것 같아요.

피드백 적용하고 미션 진행하시느라 고생 많으셨습니다!
`pixelOf` 함수에 대해 코멘트 하나 남겨두었어요. 확인 바랍니다🤤

더 진행하는 것보다 다음 미션에서 좀 더 다이브 하는 게 좋을 것 같아요.
이만 머지할게요✔

taeyoungs commented:
> 타입스크립트에는 타입 공간과 값 공간이라는 게 존재합니다. 값 공간은 자바스크립트 코드가 존재하며, 타입 공간은 타입만 위치하게 돼요. 그런데 두 곳 모두 존재하는 코드가 있습니다. typeof, class, enum 등이 대표적이에요. 더 이질적인 것은 enum 입니다. 타입스크립트에서 파생된 문법임에도 불구하고 자바스크립트 코드로 변환되는 것은... "컴파일 레벨에서 상수로 변경되는 것인데 그럴거면 enum이 아니라 상수를 쓰겠다" 입니다.
> 또한 Enum은 근본적인 문제가 있습니다. 인덱스와 키로 값에 접근할 수 있는 장점이 있지만, 강타입 언어에서 강제하는 indeOfRange 등의 에러는 타입스크립트의 Enum에 존재하지 않습니다. 이런 이유로, 저는 일절 사용하지 않아요.

자세한 답변 정말 감사합니다! 🙇‍♂️
